lib/query.js

```diff
@@ -2240,8 +2241,12 @@ Query.prototype._find = wrapThunk(function(callback) {
     if (this.options.explain) {
       return callback(null, docs);
     }
-
     if (!mongooseOptions.populate) {
+      if (mongooseOptions.lean && mongooseOptions.lean.versionKey === false && _this.schema.options.versionKey) {
+        docs.forEach((doc) => {
+          delete doc[_this.schema.options.versionKey];
```

Comment:
delete on an Object is naturally slow. Can we please first check if the key is actually set and if so then delete it?

```suggestion
          if (_this.schema.options.versionKey in doc) 
          { 
              delete doc[_this.schema.options.versionKey];
          }
```

Or better first create a local variable `cont versionKey = _this.schema.options.versionKey;` to improve the perf by avoiding the lookups.

Reply:
@Uzlopak I see this is a repetitive concern, what do you think of creating a utility function that deletes and checks before deletion? Hopefully, the v8 team will address this slow deletion issue sometime in the future and we won't need such utility.

```js
function deleteFromObject(object, fieldToDelete) { 
  if (fieldToDelete in object) {
    delete object[fieldToDelete]; 
  }
}
```

Reply:
I think a deleteFromObject function would result in a bigger deoptimization in v8, as the First Parameter would be Always polymorph and the fieldToDelete would be a byValue Parameter, duplicating the memory usage for the String. Probably would be a heavier Impact than checking If the field exists in the object and making deleting directly faster.

Maybe i should Benchmark this further. So neglect this remark for now

Reply:
But we should definitely store versionKey in a temp variable to avoid the lookup.

Reply:
Honestly, in my head, all synchronous code is free unless it becomes a problem.

So the performance of `delete` is something I wouldn't ever think about unless someone complains about it, there is generally much more room for performance optimization than the synchronous code.

But you do make a good point when it comes to hot paths where optimizing synchronous code has a noticeable effect on performance.